### PR TITLE
cluster up extended tests

### PR DESCRIPTION
**Pull request by csrwng**
Adds a set of extended tests for cluster up and fixes the following issues:
- cluster up won't start on a machine where dnsmasq is bound to port 53 on a certain interface
- on a clean Docker for Mac install, cluster up won't create directory shares and fail
- login fails on Docker for Mac when a non-numeric public hostname is specified

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees @stevekuznetsov ptal

File: test/extended/clusterup.sh [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+    echo "cluster up -- ${test}: ok"
+    os::test::junit::declare_suite_end
+}
+
+function os::test::extended::clusterup::verify_router_and_registry () {
+    os::cmd::expect_success "oc login -u system:admin"
+    os::cmd::expect_success_and_text "oc get svc -n default" "docker-registry"
+    os::cmd::expect_success_and_text "oc get svc -n default" "router"
+    os::cmd::try_until_success "oc get endpoints docker-registry -n default | grep ':5000'" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]:
```
os::cmd::try_until_text "oc get endpoints docker-registry -n default -o jsonpath='{ .subsets[*].ports[?(@.name=="5000-tcp")].port }'" "5000" "$(( 10*minute ))" 1
```

File: test/extended/clusterup.sh [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    os::cmd::try_until_success "oc get endpoints router -n default | grep ':80'" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]:
```
os::cmd::try_until_text "oc get endpoints router -n default -o jsonpath='{ .subsets[*].ports[?(@.name=="80-tcp")].port }'" "80" "$(( 10*minute ))" 1
```

File: test/extended/clusterup.sh [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    os::cmd::expect_success "oc login -u developer"
+}
+
+function os::test::extended::clusterup::verify_image_streams () {
+    os::cmd::try_until_success "oc get is -n openshift ruby -o jsonpath='{ .status.tags[*].tag }' | grep latest" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: You can grab the latest tag with `-o jsonpath='{ .status.tags[?(@.tag=="latest")] }' and maybe test with 

```
os::cmd::try_until_text "oc get is -n openshift ruby -o jsonpath='{ .status.tags[?(@.tag=="latest")].tag }'" "latest" "$(( 10*minute ))" 1
```

File: test/extended/clusterup.sh [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+function os::test::extended::clusterup::verify_ruby_build () {
+    os::cmd::expect_success "oc new-app https://github.com/openshift/ruby-hello-world.git"
+    os::cmd::try_until_success "oc get builds -l app=ruby-hello-world | grep 'Complete'" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]:
```
os::cmd::try_until_text "oc get build -l app=ruby-hello-world -o jsonpath='{ .items[*].status.phase }'" "Complete"
```

File: test/extended/clusterup.sh [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+function os::test::extended::clusterup::verify_persistent_volumes () {
+    os::cmd::expect_success "oc login -u system:admin"
+    os::cmd::expect_success_and_text "oc get jobs -n default" "persistent-volume-setup"
+    os::cmd::try_until_success "oc get job persistent-volume-setup -n default -o jsonpath='{ .status.succeeded }' | grep 1" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: use `try_until_text` not `grep`

File: test/extended/clusterup.sh [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    os::test::extended::clusterup::verify_router_and_registry
+    os::test::extended::clusterup::verify_image_streams
+    os::test::extended::clusterup::verify_ruby_build
+    os::test::extended::clusterup::verify_persistent_volumes
+}
+
+
+# Test with host dirs specified
+function os::test::extended::clusterup::internal::hostdirs () {
+    local data_dir="$(mktemp -d)"
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: don't mix scoping and subshells
```sh
local data_dir; data_dir="$( mktemp -d )"
```

Also root these somewhere under `$BASETMPDIR`

File: test/extended/clusterup.sh [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    fi
+    TMPDIR="${tmp}" os::test::extended::clusterup::internal::hostdirs
+}
+
+# Tests creating a cluster with an alternate image and tag
+function os::test::extended::clusterup::image () {
+    os::test::extended::clusterup::standard_test \
+        --image="registry.access.redhat.com/openshift3/ose" \
+        --version="v3.3"
+    os::cmd::expect_success_and_text "docker inspect origin" "registry.access.redhat.com/openshift3/ose:v3.3"
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: Should use `--format` here to grab the field you want this to appear in

File: test/extended/clusterup.sh [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    os::cmd::expect_success "oc cluster down"
+}
+
+# Tests creating a cluster with a public hostname
+function os::test::extended::clusterup::publichostname () {
+    os::test::extended::clusterup::standard_test \
+        --public-hostname="my.server.name"
+    os::cmd::expect_success "docker exec origin cat /var/lib/origin/openshift.local.config/master/master-config.yaml | grep 'masterPublicURL.*my\.server\.name'"
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: `expect_success_and_text` not `grep`

File: test/extended/clusterup.sh [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    os::cmd::expect_success "oc cluster down"
+    docker rm -f port53
+}
+
+# If an argument was passed, run the test that was specified
+if [[ "${1}" != "" ]]; then
+    os::test::extended::clusterup::run_test "${1}"
+    exit
+fi
+
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: I would make this part look like


```sh
readonly default_tests=(
	"noargs"
	"hostdirs"
	"image"
	"publichostname"
	"numerichostname"
	"metrics"
	"logging"
	"svcaccess"
	"portinuse"
)

tests=("${1:-"${default_tests[@]}"}")

for test in "${tests[@]}"; do
    if ! os::test::extended::clusterup::run_test "${test}"; then
        failed=1
    fi
done

if [[ -n "${failed:-}" ]]; then
    os::log::fatal "Cluster up tests failed!"
fi
```

File: test/extended/clusterup.sh [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    os::cmd::expect_success "timeout 2m oc rsh dc/git curl http://git:8080/_/healthz"
+    os::cmd::expect_success "oc cluster down"
+}
+
+# Verifies that cluster up can start when a process is bound
+# to the loopback interface with tcp port 53
+function os::test::extended::clusterup::portinuse () {
+    # Start listening on the host's 127.0.0.1 interface, port 53
+    docker run -d --name=port53 --entrypoint=/bin/bash --net=host openshift/origin-gitserver:latest -c "socat TCP-LISTEN:53,bind=127.0.0.1,fork SYSTEM:'echo hello'"
+    os::cmd::expect_success "oc cluster up"
```
Comment (bparees) [inline test/extended/clusterup.sh]: shouldn't you do some verification of the cluster?

File: test/extended/clusterup.sh [hunk 11]
```diff
@@ -0,0 +1,194 @@
+#!/bin/bash
+source "$(dirname "${BASH_SOURCE}")/../../hack/lib/init.sh"
+trap os::test::junit::reconcile_output EXIT
```
Comment (bparees) [inline test/extended/clusterup.sh]: is anything going to bring down the cluster on failure?

Comment (csrwng) [inline test/extended/clusterup.sh]: working on that

File: test/extended/clusterup.sh (hunk 10 above)
Comment (csrwng) [inline test/extended/clusterup.sh]: so this is a test for a very specific situation. When you're running dnsmasq, cluster-up fails to start. I don't think we need to make the test longer. If it starts successfully, it means we don't have the bug.

Comment (bparees) [inline test/extended/clusterup.sh]: ok

File: test/extended/clusterup.sh [hunk 12]
```diff
@@ -0,0 +1,194 @@
+#!/bin/bash
```
Comment (csrwng) [inline test/extended/clusterup.sh]: yes, I would like to run this extended job on PRs with the default extended test annotation for rosiebot

File: test/extended/clusterup.sh [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+source "$(dirname "${BASH_SOURCE}")/../../hack/lib/init.sh"
+trap os::test::junit::reconcile_output EXIT
+
+
+os::util::ensure::built_binary_exists 'oc'
+export PATH="$(dirname "$( os::util::find::built_binary oc )")":"${PATH}"
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: We actually call that in `hack/lib/init.sh`

**Comment by csrwng** [comment]
Added cleanup code.
@stevekuznetsov it should be ready for a final review.

File: test/extended/clusterup.sh [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+function os::test::extended::clusterup::run_test () {
+    local test="${1}"
+    local funcname="os::test::extended::clusterup::${test}"
+    local global_home="${HOME}"
+    os::log::info "starting test -- ${test}"
+
+    os::test::junit::declare_suite_start "extended/clusterup/${test}"
+
+	local test_home="${ARTIFACT_DIR}/${test}/home"
+	mkdir -p "${test_home}"
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: indentation nit

File: test/extended/clusterup.sh [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    export HOME="${test_home}"
+    pushd "${HOME}" &> /dev/null
+    os::log::info "Using ${HOME} as home directory"
+	${funcname}
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: indentation nit

File: test/extended/clusterup.sh [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+
+    os::log::info "cluster up -- ${test}: ok"
+    os::test::junit::declare_suite_end
+}
+
+function os::test::extended::clusterup::verify_router_and_registry () {
+    os::cmd::expect_success "oc login -u system:admin"
+    os::cmd::expect_success_and_text "oc get svc -n default" "docker-registry"
+    os::cmd::expect_success_and_text "oc get svc -n default" "router"
+    os::cmd::try_until_text "oc get endpoints docker-registry -n default" ":5000" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: `jsonpath`

File: test/extended/clusterup.sh [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+    os::cmd::try_until_text "oc get endpoints router -n default" ":80" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: `jsonpath`

File: test/extended/clusterup.sh [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+    os::cmd::try_until_text "oc get job persistent-volume-setup -n default -o jsonpath='{ .status.succeeded }'" "1" $(( 10*minute )) 1
+    os::cmd::expect_success "oc get pv/pv0100"
+    os::cmd::expect_success "oc login -u developer"
+}
+
+function os::test::extended::clusterup::verify_metrics () {
+    os::cmd::expect_success "oc login -u system:admin"
+    os::cmd::expect_success_and_text "oc get pods -n openshift-infra" "metrics-deployer"
+    os::cmd::try_until_text "oc get pods -n openshift-infra -l job-name=metrics-deployer-pod" "Completed" $(( 10*minute )) 1
+	os::cmd::try_until_text "oc get endpoints -n openshift-infra" ":3000" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: `jsonpath`

Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: indentation nit

File: test/extended/clusterup.sh [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+    os::cmd::expect_success "oc login -u developer"
+}
+
+function os::test::extended::clusterup::verify_logging () {
+    os::cmd::expect_success "oc login -u system:admin"
+    os::cmd::expect_success_and_text "oc get pods -n logging" "logging-deployer"
+    os::cmd::try_until_text "oc get pods -n logging -l logging-infra=deployer" "Completed" $(( 10*minute )) 1
+    os::cmd::try_until_text "oc get endpoints logging-kibana -n logging" ":3000" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: `jsonpath`

File: test/extended/clusterup.sh [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+
+# Test with host dirs specified
+function os::test::extended::clusterup::internal::hostdirs () {
+    local data_dir="${BASE_DIR}/data"
+    local config_dir="${BASE_DIR}/config"
+    local pv_dir="${BASE_DIR}/pv"
+    local volumes_dir="${BASE_DIR}/volumes"
+	os::cmd::expect_success "mkdir -p ${data_dir} ${config_dir} ${pv_dir} ${volumes_dir}"
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: indentation nit

File: test/extended/clusterup.sh [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+    os::test::extended::clusterup::standard_test --logging
+    os::test::extended::clusterup::verify_logging
+}
+
+# Verifies that a service can be accessed by a peer pod
+# and by the pod running the service
+function os::test::extended::clusterup::svcaccess () {
+    os::cmd::expect_success "oc cluster up"
+    os::cmd::expect_success "oc create -f ${OS_ROOT}/examples/gitserver/gitserver-persistent.yaml"
+    os::cmd::try_until_text "oc get endpoints git" ":8080" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: `jsonpath`

File: test/extended/clusterup.sh [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+	"metrics"
+	"logging"
+)
+
+tests=("${1:-"${default_tests[@]}"}")
+
+for test in "${tests[@]}"; do
+	cleanup_func=$("os::text::extended::clusterup::cleanup_func" "${test}")
+	trap "${cleanup_func}" EXIT
+    os::test::extended::clusterup::run_test "${test}"
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: indentation nit

File: test/extended/clusterup.sh [hunk 23]
```diff
@@ -0,0 +1,215 @@
+#!/bin/bash
+source "$(dirname "${BASH_SOURCE}")/../../hack/lib/init.sh"
+trap os::test::junit::reconcile_output EXIT
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: i have not had time to factor this out yet, so you will need to cargo-cult the following for now:

https://github.com/openshift/origin/blob/master/test/extended/gssapi.sh#L21-L24
https://github.com/openshift/origin/blob/master/test/extended/gssapi.sh#L38-L55

**Comment by csrwng** [comment]
@stevekuznetsov comments addressed

File: test/extended/clusterup.sh [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+
+    os::log::info "cluster up -- ${test}: ok"
+    os::test::junit::declare_suite_end
+}
+
+function os::test::extended::clusterup::verify_router_and_registry () {
+    os::cmd::expect_success "oc login -u system:admin"
+    os::cmd::expect_success_and_text "oc get svc -n default" "docker-registry"
+    os::cmd::expect_success_and_text "oc get svc -n default" "router"
+    os::cmd::try_until_text "oc get endpoints docker-registry -o jsonpath='{ .subsets[*].ports[*].port }' -n default" "5000" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: This has the chance of matching port `50000` and `15000` right? That's why the `jsonpath` I suggested grabbed a specific port entry

File: test/extended/clusterup.sh [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+    os::cmd::expect_success "oc login -u developer"
+}
+
+function os::test::extended::clusterup::verify_image_streams () {
+    os::cmd::try_until_text "oc get is -n openshift ruby -o jsonpath='{ .status.tags[*].tag }'" "latest" $(( 10*minute )) 1
+}
+
+function os::test::extended::clusterup::verify_ruby_build () {
+    os::cmd::expect_success "oc new-app https://github.com/openshift/ruby-hello-world.git"
+    os::cmd::try_until_text "oc get builds -l app=ruby-hello-world" "Complete" $(( 10*minute )) 1
```
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: `jsonpath`

File: test/extended/clusterup.sh (hunk 24 above)
Comment (csrwng) [inline test/extended/clusterup.sh]: Sorry I didn't see any `jsonpath` suggestion, but I will update.

**Comment by csrwng** [comment]
addressed

File: test/extended/clusterup.sh (hunk 24 above)
Comment (stevekuznetsov) [inline test/extended/clusterup.sh]: It was in the first review

**Comment by csrwng** [comment]
@stevekuznetsov this will not go green right now because we're missing images for metrics and logging. 

@smarterclayton ^^ ? 

[testextended][extended:clusterup]

**Comment by csrwng** [comment]
@stevekuznetsov I'm going to split this up into fixes for cluster up and extended tests so we can at least get the fixes in.

**Comment by smarterclayton** [comment]
Trying to get to this today

**Comment by csrwng** [comment]
[testextended][extended:clusterup]

**Comment by csrwng** [comment]
@stevekuznetsov can you tag this pr to run extended tests again?

**Comment by stevekuznetsov** [comment]
Looks like a code problem:
```
Running TEST_REPORT_DIR=/tmp/openshift/test-extended/junit OPENSHIFT_TEST_IMAGE_REGISTRY=ci.dev.openshift.redhat.com:5000 JUNIT_REPORT=true test/extended/clusterup.sh...
test/extended/clusterup.sh: line 6: LOG_DIR: unbound variable
```

**Comment by csrwng** [comment]
The metrics test is failing because pulling images is timing out:
```
E0213 21:44:24.793905   16043 kube_docker_client.go:324] Cancel pulling image "openshift/origin-metrics-hawkular-metrics:v1.5.0-alpha.2" because of no progress for 1m0s
, latest progress: "9905334bc95a: Extracting [==================================================>] 68.86 MB/68.86 MB"
E0213 21:44:24.794123   16043 docker_manager.go:2324] container start failed: ErrImagePull: net/http: request canceled
E0213 21:44:24.794254   16043 pod_workers.go:184] Error syncing pod 5f6882f7-f235-11e6-a9b8-0e371b4e866c, skipping: failed to "StartContainer" for "hawkular-metrics" with ErrImagePull: "net/http: request canceled"
```

@stevekuznetsov have you seen this elsewhere on a Jenkins test?
The pull works fine locally.

**Comment by csrwng** [comment]
It looks like the images are eventually pulled, but the job to setup metrics has a deadline of 5min, so it stops trying after that. I will extend the deadline to 10min and see how it does.

**Comment by stevekuznetsov** [comment]
I have not seen that before, no. It's a relatively large amount of data, right? Doesn't seem unreasonable that it would flake out.

**Comment by csrwng** [comment]
Included in https://github.com/openshift/origin/pull/14729

**Pull request closed without merging**